### PR TITLE
Fix form reset for uncontrolled select elements

### PR DESCRIPTION
**Pull request by aweary**
Resolves https://github.com/facebook/react/issues/11010

`defaultSelected` will now be set on the select's option's elements based on the value(s) in `defaultValue`. Added a test fixture, too.

### CODE REVIEW

File: src/renderers/dom/fiber/wrappers/ReactDOMFiberSelect.js [hunk 1]
```diff
@@ -103,6 +107,9 @@ function updateOptions(
     for (let i = 0; i < options.length; i++) {
       if (options[i].value === selectedValue) {
         options[i].selected = true;
+        if (setDefaultSelected) {
+          options[i].defaultSelected = true;
```
Comment (jquense) [inline src/renderers/dom/fiber/wrappers/ReactDOMFiberSelect.js]: perhaps this should be set once on mount? That is a bit iffy i guess. Ostensibly `defaultValue` is set once for inputs and changing it after that is a noop for the life of the component. It seems like now tho you can set it, change it and reset the form and it'll use the new value. That may not be worth the complexity to address, but it opens up a new inconsistency perhaps.

Comment (aweary) [reply src/renderers/dom/fiber/wrappers/ReactDOMFiberSelect.js]: > Ostensibly defaultValue is set once for inputs and changing it after that is a noop for the life of the component.

That isn't the case, as far as I've seen. [Here's an example](https://jsfiddle.net/gocubxhk/) that you can update `defaultValue` and a form reset will reset to the new `defaultValue` option. That happens in `ReactDOMFiberInput.updateWrapper` [here](https://github.com/facebook/react/blob/master/src/renderers/dom/fiber/wrappers/ReactDOMFiberInput.js#L220-L232). In both cases, updating `.defaultValue` is guarded by a loose `null` check for `value` first, so I think it should be consistent?

Comment (jquense) [reply src/renderers/dom/fiber/wrappers/ReactDOMFiberSelect.js]: ah ok, so it's consistent with the other input reset behavior sorry. carry on then :P

**Comment by jquense** [comment]
Do we know yet if `defaultSelected` works for all browsers on reset or do we still need to do some testing?

**Comment by aweary** [comment]
@jquense I tested in IE9, and the latest Chrome, Firefox, and Safari on macOS. It would be great if we could run through the fixtures in some browsers.

You can find them here:  http://uncontrolled-select-reset.surge.sh/

> Just to be sure: does form reset trigger a change event on selects? It should not, at least it does not for inputs last i checked, but i don't believe in consistent DOM APIs anymore

@nhunzaker from what I saw, it doesn't trigger a change event, so it should be consistent.

**Comment by jquense** [comment]
I ran it through on android 4.4, old IOS, edge and safari * and 👍

**Comment by nhunzaker** [comment]
I can do some extra testing in an hour or so. 

We should figure out how to map reduce this via a plea over Twitter :)

**Comment by nhunzaker** [comment]
Finally got to testing. Tried to fill in the gaps:

- IE10, IE11, IE Edge 15
- Opera 18
- Safari 11
- iOS Safari 10.3
- Current Chrome for Android

**Comment by gaearon** [comment]
Anything blocking this?

**Comment by nhunzaker** [comment]
Nothing that I see. @aweary this is checking out for me everywhere. Any final concerns before merging?

**Comment by aweary** [comment]
@gaearon had mentioned that @sebmarkbage might have some thoughts on this, so if it looks good to him then 🚀

**Pull request merged**